docs/best-practices/gitops-with-kruise.md

```diff
@@ -0,0 +1,183 @@
+For example, there exists **dev_branch** and **master** corresponding to the git configuration for test and prd environments, as follows:
+
+![git_envs](/img/docs/best-practices/git_envs.png)
+
+### Configuring Argo Application, And Deploying Guestbook
+1. Argo-cd provides a [Declarative Setup](https://argo-cd.readthedocs.io/en/stable/operator-manual/declarative-setup/), as follows:
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  ## Represents dev env deployment configuration
+  name: guestbook-dev
+  namespace: argocd
+spec:
+  project: default
+  source:
+    repoURL: https://github.com/zmberg/samples.git
+    ## dev branch
+    targetRevision: dev_branch
+    path: guestbook
+  destination:
+    server: https://kubernetes.default.svc
+    ## deploy in default namespace
+    namespace: default
+```
+
+2. Create the Application CRD resource into k8s cluster, and complete the deployment using the argocd command line tool,
+refer to the [documentation](https://argo-cd.readthedocs.io/en/stable/getting_started/), as follows:
+
+![guestbook](/img/docs/best-practices/argo_sync_unknown.png)
+
+**Note: After publishing through argo-cd, we found a very strange problem. argo-cd actually did not determine whether the cloneSet publishing is complete and successful,
+and then judged that this Application publishing has been Healthy, which is a very serious risk point.**
+
+### Configuring CloneSet Custom CRD Argo-cd Resource Health
+
+By reading argo-cd source code, we found that argo has built-in health check for k8s native resources, such as Deployment, Service, Ingress and so on. But the custom resources for Kruise CloneSet extensions are ignored directly. Is there no solution to this problem? Do we have to modify the source code of argo-cd?
+
+**After reviewing the argo-cd documentation in detail, I found out that argo-cd extends the [Custom CRD Health Checks](https://argo-cd.readthedocs.io/en/stable/operator-manual/health/#custom-health-checks) method for the CloneSet custom CRD.**
+Configure the *argocd-cm Configmap* resource according to the documentation, as follows:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  labels:
+    app.kubernetes.io/name: argocd-cm
+    app.kubernetes.io/part-of: argocd
+  name: argocd-cm
+  namespace: argocd
+data:
+  resource.customizations.health.apps.kruise.io_CloneSet: |
+
+    hs = {}
+    -- if paused
+    if obj.spec.updateStrategy.paused then
+      hs.status = "Suspended"
+      hs.message = "CloneSet is Suspended"
+      return hs
+    end
+
+    -- check cloneSet status
+    if obj.status ~= nil then
+      if obj.status.observedGeneration < obj.metadata.generation then
+        hs.status = "Progressing"
+        hs.message = "Waiting for rollout to finish: observed cloneSet generation less then desired generation"
+        return hs
+      end
+
+      if obj.status.updatedReplicas < obj.spec.replicas then
```

Comment:
if updatedreadyreplicas+partition = replicas ，status should also be "Suspended"

Reply:
It is estimated to be difficult to configure the partition using gitops, so the partition is not considered at this stage